src/MotorPlugin.cc

```diff
@@ -0,0 +1,704 @@
+    }
+  }
+
+  private: gz::transport::parameters::ParametersRegistry * registry{nullptr};
+  private: T * obj{nullptr};
+  private: std::function<double(const T&)> getter;
+  private: std::function<void(T&, double)> setter;
+  private: std::string_view prefix;
+  private: std::string_view name;
+  private: std::string scopedName;
+};
+
+//////////////////////////////////////////////////
+class MotorPlugin::Impl
+{ 
+  //
+  public: void OnPwmMsg(const google::protobuf::Message &_msg,
+                        const gz::transport::MessageInfo &_info);
+  
+                        //! Helper to initialise and declare a PID parameter
+  public: template<typename Getter, typename Setter>
+  void DeclareParameter(
+      ParameterProxy<math::PID>& param,
+      Getter&& getter,
+      Setter&& setter,
+      const std::string& prefix)
+  {
+    param.Init(registry, &posPid,
+      std::forward<Getter>(getter),
+      std::forward<Setter>(setter),
+      prefix);
+    param.Declare();
+  }
+  /// \brief World occupied by the parent model.
+  public: World world{kNullEntity};
+  
+  /// \brief Name of the world entity.
+  public: std::string worldName;
+
+  /// \brief Model entity of Motor Model.
+  public: Model parentModel{kNullEntity};
+
+  /// \brief Name of the model entity.
+  public: std::string parentModelName;
+
+  /// \brief Array of controllers
+  public: std::vector<Control> controls;
+
+  /// \brief Array of pwm command topics
+  public: std::vector<std::string> topics;
+  
+  /// \brief keep track of controller update sim-time.
+  public: std::chrono::steady_clock::duration lastMotorModelUpdateTime{0};
+
+  ///
+  public: std::string joint_name_;
+  
+  ///
+  public: gz::sim::Entity joint_entity_ = gz::sim::kNullEntity;
+
+  /// \brief Array of subscribers for PWM topics
+  public: std::vector<gz::transport::Node::Subscriber> subscribers;
+  
+  /// \brief Store PWM values indexed by channel
+  public: std::map<std::string, std::string> pwmValues;
+  
+  /// \brief Mutex to protect PWM values
+  public: std::mutex pwmMutex;
+
+  //
+  public: bool validConfig{false};
+
+  //
+  gz::transport::Node node;
+
+  /// \brief Joint Entity
+  public: Joint joint{kNullEntity};
+
+   /// \brief Position PID controller.
+  public: math::PID posPid;
+
+  /// \brief Parameters registry
+  public: transport::parameters::ParametersRegistry * registry;
+
+  /// Dynamic parameters
+  public: ParameterProxy<math::PID> pGain{"p_gain"};
+  public: ParameterProxy<math::PID> iGain{"i_gain"};
+  public: ParameterProxy<math::PID> dGain{"d_gain"};
+  public: ParameterProxy<math::PID> iMax{"i_max"};
+  public: ParameterProxy<math::PID> iMin{"i_min"};
+  public: ParameterProxy<math::PID> cmdMax{"cmd_max"};
+  public: ParameterProxy<math::PID> cmdMin{"cmd_min"};
+  public: ParameterProxy<math::PID> cmdOffset{"cmd_offset"};
+
+};
+
+//////////////////////////////////////////////////
+void MotorPlugin::Impl::OnPwmMsg(const google::protobuf::Message &_msg,
```

Comment:
Argument can be a `const msgs::Double &_msg` as this is what is expected. Change declaration and

```c++
#include <gz/msgs/double.pb.h>
```

rather than

```c++
#include <google/protobuf/message.h>
```